File: app/lib/features/highlight/domain/highlight.dart
import 'package:blooms/features/highlight/domain/highlight_content.dart';
import 'package:blooms/features/highlight/domain/highlight_period.dart';
import 'package:blooms/features/highlight/domain/highlight_state.dart';
import 'package:blooms/gen/strings.g.dart';
import 'package:blooms/theme/my_date_format.dart';
import 'package:blooms/utils/timestamp_converter.dart';
import 'package:blooms/utils/typedefs.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'highlight.freezed.dart';
part 'highlight.g.dart';

@freezed
abstract class Highlight with _$Highlight {
  const factory Highlight({
    /// ハイライトの対象者のUID
    required String subjectUid,

    /// 内容
    required HighlightContent content,
    @TimestampConverter() Timestamp? createdAt,
    @TimestampConverter() Timestamp? updatedAt,
    @TimestampConverter() Timestamp? deletedAt,
  }) = _Highlight;

  factory Highlight.summary({
    required HighlightPeriod period,
    required String subjectUid,
    required Timestamp startAt,
  }) => Highlight(
    subjectUid: subjectUid,
    content: HighlightContentSummary(
      startAt: startAt,
      period: period,
      summary: '',
      abstract: '',
    ),
  );

  const Highlight._();

  factory Highlight.fromJson(Json json) => _$HighlightFromJson(json);

  static FromFirestore<Highlight> get fromFirestore =>
      (snapshot, _) => Highlight.fromJson(snapshot.data() ?? {});

  static ToFirestore<Highlight> get toFirestore =>
      (data, _) => TimestampConverter.updateServerTimestamp(data.toJson());

  HighlightRange get highlightRange {
    final (startAt, period) = switch (content) {
      HighlightContentSummary(:final startAt, :final period) => (
        startAt,
        period,
      ),
      HighlightContentEmpty() => throw UnimplementedError(),
    };

    final startDate = myDateFormat(
      startAt.toDate().subtract(Duration(days: period.days)),
    );
    final endDate = myDateFormat(startAt.toDate());

    return (startDate: startDate, endDate: endDate);
  }

  /// ハイライトの対象期間を表す文字列
  ///
  /// e.g.
  /// 2月1日(土)
  /// 1月25日(土) から 2月1日(土)
  String get highlightRangeString {
    final (:startDate, :endDate) = highlightRange;

    if (period == HighlightPeriod.past1day) {
      return endDate;
    }

    return i18n.highlight.xToY(x: startDate, y: endDate);
  }

  HighlightPeriod? get period {
    return switch (content) {
      HighlightContentSummary(:final period) => period,
      HighlightContentEmpty() => null,
    };
  }

  HighlightState? get state {
    return switch (content) {
      HighlightContentSummary(:final state) => state,
      HighlightContentEmpty() => null,
    };
  }
}

typedef HighlightRange = ({String startDate, String endDate});

File: firebase/functions/src/features/highlight/domain/highlight.ts
import { FieldValue, FirestoreDataConverter, Timestamp } from 'firebase-admin/firestore';
import { logger } from 'firebase-functions';

export type HighlightPeriod = 'past1day' | 'past7days' | 'past14days' | 'past21days' | 'past28days';
export type HighlightType = 'summary' | 'empty';
export type HighlightState = 'pending' | 'inProgress' | 'success' | 'failure';
export type HighlightContentInterfaces = HighlightContentSummary | HighlightContentEmpty;

export interface HighlightContentSummary {
  type: 'summary';
  startAt: Timestamp;
  period: HighlightPeriod;
  summary: string;
  abstract: string;
  state: HighlightState;
  promptFileUri: string | null;
}

export interface HighlightContentEmpty {
  type: 'empty';
}

export class Highlight {
  constructor(
    readonly createdAt: Timestamp | FieldValue,
    readonly updatedAt: Timestamp | FieldValue,
    readonly deletedAt: Timestamp | null,
    readonly subjectUid: string,
    readonly content: HighlightContentInterfaces,
  ) { }
}

export const highlightConverter: FirestoreDataConverter<Highlight> = {
  toFirestore: (highlight: Highlight) => ({
    createdAt: highlight.createdAt,
    updatedAt: highlight.updatedAt,
    deletedAt: highlight.deletedAt,
    subjectUid: highlight.subjectUid,
    content: highlight.content,
  }),
  fromFirestore: snapshot => new Highlight(
    snapshot.get('createdAt') as Timestamp,
    snapshot.get('updatedAt') as Timestamp,
    snapshot.get('deletedAt') as Timestamp | null,
    snapshot.get('subjectUid') as string,
    highlightContentConverter(snapshot.get('content') as Record<string, unknown>),
  ),
};

function highlightContentConverter(value: Record<string, unknown>): HighlightContentSummary | HighlightContentEmpty {
  if (value == null) {
    return {
      type: 'empty',
    };
  }

  const type = value['type'] as HighlightType | null;

  switch (type) {
    case 'summary':
      return {
        type: type,
        startAt: value['startAt'] as Timestamp,
        period: value['period'] as HighlightPeriod,
        summary: value['summary'] as string,
        abstract: value['abstract'] as string,
        state: value['state'] as HighlightState,
        promptFileUri: value['promptFileUri'] as string | null,
      };
    default:
      logger.error(`Invalid type: ${type}`);
      return {
        type: 'empty',
      };
  }
}
